packages/commonwealth/server/util/requirementsModule/validateTopicGroupsMembership.ts

```diff
@@ -0,0 +1,80 @@
+import { DB } from '../../models';
+import { TokenBalanceCache } from '../../../../token-balance-cache/src';
+import { ChainInstance } from '../../models/chain';
+import { AddressInstance } from '../../models/address';
+import { FEATURE_FLAG_GROUP_CHECK_ENABLED } from '../../config';
+import validateGroupMembership from './validateGroupMembership';
+import validateTopicThreshold from '../validateTopicThreshold';
+import { ServerError } from '../../../../common-common/src/errors';
+import { Op } from 'sequelize';
+
+export const Errors = {
+  InsufficientTokenBalance: 'Insufficient token balance',
+  BalanceCheckFailed: 'Could not verify user token balance',
+};
+
+/**
+ * Validates if a given user address passes a set of requirements and grants access for
+ * all groups of the given topic. Depending on the FEATURE_FLAG_GROUP_CHECK_ENABLED
+ * feature flag, may use Gating API implementation or original TBC implementation.
+ * @param models DB handle
+ * @param tokenBalanceCache Token balance cache handle
+ * @param topicId ID of the topic
+ * @param chain Chain of the groups
+ * @param address Address to check against requirements
+ * @returns validity with optional error message
+ */
+export async function validateTopicGroupsMembership(
+  models: DB,
+  tokenBalanceCache: TokenBalanceCache,
+  topicId: number,
+  chain: ChainInstance,
+  address: AddressInstance
+): Promise<{ isValid: boolean; message?: string }> {
+  if (FEATURE_FLAG_GROUP_CHECK_ENABLED) {
+    // check via groups
+
+    // get all groups of topic
+    const topic = await models.Topic.findOne({
```

Comment:
This might not work, but could we add some sql magic to join this with the group query? Something like:

const topic = await models.Topic.findOne({
  where: {
    chain_id: chain.id,
    id: topicId,
    attributes: ['requirements']

  },
  include: [
    {
      model: models.Group,
      where: {
        id: { [Op.in]: Sequelize.literal('(topic.group_ids)') },
      },
     attributes: ['requirements']
    },
  ],
});

I am not 100% the sequelize will work, but I am trying to get at this query:
SELECT g.requirements
FROM "Topics" as t
JOIN "Group" as g ON g.id IN (t.group_ids)

Which will effectively join the group

Reply:
Doesn't work since `group_ids` is an array. Sequelize only supports many-many relationships via a through table

Reply:
Alright yeah– tried the code and sequelize complains about a lack of association. But an association doesn't work with the `group_ids` array– only with a direct fkey association or a through table.

Kinda suggests that maybe the DB should be restructured, but that's for another day 😅